### ISSUE TITLE
ASP.NET Core 2.1: An existing connection was forcibly closed by the remote host

### ISSUE THREAD

**Issue by ufalak** [aspnet/KestrelHttpServer#2796]
#2554

After migrating to the last version of  asp net core 2.1 i start getting this exception:

```
An existing connection was forcibly closed by the remote host
Source:System.Private.CoreLib
InnerException:System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitable.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
StackTrace:   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.ReadAsync(CancellationToken token)
   at System.IO.Pipelines.Pipe.DefaultPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.PumpAsync()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.ReadAsync(CancellationToken token)
   at System.IO.Pipelines.Pipe.DefaultPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Formatters.JsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextExceptionFilterAsync() 
2018-08-08 11:49:39.2834|7-5000075-17ef1173-99f0-e8cb-299d-ca9d9791f087|Catalogo|ERROR|Mensaje:An existing connection was forcibly closed by the remote host
Source:System.Private.CoreLib
InnerException:System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitable.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
StackTrace:   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.ReadAsync(CancellationToken token)
   at System.IO.Pipelines.Pipe.DefaultPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.PumpAsync()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.ReadAsync(CancellationToken token)
   at System.IO.Pipelines.Pipe.DefaultPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Formatters.JsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextExceptionFilterAsync() 
```

**Comment by muratg** [comment]
@ufalak Looks like it's a benign client discosnnect. Is the issue that you started seeing this, or are there actual issues you're hitting?

It would be interesting to see this from actual logs (looks like you copied the stack traces from elsewhere.)

**Comment by ufalak** [comment]
In fact, my principal problem is that i don really know what effect is taking for the client because i cant reproduce it. 
I dont know wich conection its making closed and neither in wich condition and neither whats effect take to the client.

This is extracted from the log in production environment and i started to see this when i move to net core 2.1.

**Comment by ufalak** [comment]
Doesnt look that is the same error, because i dont get that much nested exceptions.
I need to know:
1) what exactly its throwing this exception of "An existing connection was forcibly closed by the remote host
Source:System.Private.CoreLib
InnerException:System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host" 

2) What effect its causing to the user that is navigating the site at that moment?
3) how can i catch and take any action on this exception?

Thanks!

**Comment by halter73** [comment]
> 1. what exactly its throwing this exception of "An existing connection was forcibly closed by the remote host
Source:System.Private.CoreLib
InnerException:System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host"

An internal call by the server to [Socket.ReceiveAsync()](https://msdn.microsoft.com/en-us/library/system.net.sockets.socket.receiveasync%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396).

> 2. What effect its causing to the user that is navigating the site at that moment?

Hopefully very little, because it's the client (e.g. browser, googlebot, etc...) that closed the connection while the JsonInputFormatter is in the middle of parsing the request body. This could happen because the user navigating the site closed the tab or browser, an ajax request was canceled programmatically, or many other reasons. This does mean that whatever controller action the upload was supposed to trigger never executes however.

> 3. how can i catch and take any action on this exception?

It looks like you already are catching and logging the exception. The following

> 2018-08-10 00:04:27.4154|2-18000120-d59dc782-36d1-c8e0-46a2-2763e7e93d6d|Catalogo|ERROR|Mensaje:An existing connection was forcibly closed by the remote host

doesn't look like an ASP.NET log. The log source "Catalogo" isn't an ASP.NET namespace. You'll notice that the other logs include a namespace like "Microsoft.AspNetCore.Server.Kestrel" in the same place "Catalogo" appears in the above log. Those logs are logged directly by Kestrel with a relatively low "INFO" severity.

**Comment by ufalak** [comment]
Thanks so much for your answer!
And, its a core bug that this its being throwed like and exception? Are you planing to change this to not to be and exception?

**Comment by halter73** [comment]
There will always be a so-called "first chance exception" for a connection reset because of the way the manage [Socket](https://msdn.microsoft.com/en-us/library/system.net.sockets.socket%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396) API works which is used by Kestrel's new default Socket transport [introduced in 2.1.0](https://github.com/aspnet/Announcements/issues/296).

Additionally, an incomplete request body will always result in an exception being thrown from HttpContext.Request.Body.ReadAsync() regardless of whether there was an underlying connection reset, because that's the way the server indicates to the app that the request body upload did not complete successfully.

**Comment by ufalak** [comment]
Thanks again for your answer ! 
Yes, you are right, im catching the exception in global exception handler, but aprentlly this exception its something like "low importance".

So, there is any way that i can programatically distinguish wich this kind of "low important" exception to doesnt log it? 
Because, if im understanding fine, i didnt need to take any action if i get this particularry socket exception.
(i supossed that not all the socketexception are "low importance" kind of).

There is any way to reproduce this  "first chance exception" of a connection reset to test it?

**Comment by nukec** [comment]
I have same issue, not sure how to debug this:
```
System.IO.IOException:
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException (System.IO.Pipelines, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51)
   at System.IO.Pipelines.Pipe.GetFlushResult (System.IO.Pipelines, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51)
   at System.IO.Pipelines.Pipe.FlushAsync (System.IO.Pipelines, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51)
   at System.IO.Pipelines.Pipe+DefaultPipeWriter.FlushAsync (System.IO.Pipelines, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1OutputProducer.WriteAsync (Microsoft.AspNetCore.Server.Kestrel.Core, Version=2.1.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1OutputProducer.WriteDataAsync (Microsoft.AspNetCore.Server.Kestrel.Core, Version=2.1.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.WriteAsync (Microsoft.AspNetCore.Server.Kestrel.Core, Version=2.1.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpResponseStream.WriteAsync (Microsoft.AspNetCore.Server.Kestrel.Core, Version=2.1.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.AspNetCore.Http.Extensions.StreamCopyOperation+<CopyToAsync>d__2.MoveNext (Microsoft.AspNetCore.Http.Extensions, Version=2.1.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.StaticFiles.StaticFileContext+<SendAsync>d__49.MoveNext (Microsoft.AspNetCore.StaticFiles, Version=2.1.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware+<Invoke>d__7.MoveNext (Microsoft.AspNetCore.StaticFiles, Version=2.1.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware+<Invoke>d__6.MoveNext (Microsoft.AspNetCore.Diagnostics, Version=2.1.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
Inner exception System.Net.Sockets.SocketException handled at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw:
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitable.GetResult (Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets, Version=2.1.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection+<ProcessSends>d__26.MoveNext (Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets, Version=2.1.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection+<DoSend>d__25.MoveNext (Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets, Version=2.1.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60)
```

**Comment by halter73** [comment]
@nukec Are you seeing any issues other than the exception? Like I said in my last comment, here will likely always be a so-called "first chance exception" for a connection reset.

**Comment by ufalak** [comment]
The issue is that its an unhandled scenario. 
I need to be able to reproduce this exception and handle it.
This wasnt happen before the upgrade to net core 2.1 so i supposed its something new on this upgrade

**Comment by halter73** [comment]
But it looks like the exception is being handled. It's being thrown through `Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware+<Invoke>d__7.MoveNext` [here](https://github.com/aspnet/StaticFiles/blob/master/src/Microsoft.AspNetCore.StaticFiles/StaticFileMiddleware.cs#L113). Even if no other middleware handles this exception, [Kestrel does](https://github.com/aspnet/KestrelHttpServer/blob/2.1.3/src/Kestrel.Core/Internal/Http/HttpProtocol.cs#L548-L571).

**Comment by ufalak** [comment]
Yes, but its internally handled and i only handle this exception in my code as an "exception".
I need two things: 
1. how to exactly reproduce this exception 
2. how to exactly catch this particullary exception (not all the SocketException, just this one thrown from this place) to stop logging it because is aparentlly useless

**Comment by ufalak** [comment]
I write that before in [my first comment](https://github.com/aspnet/KestrelHttpServer/issues/2796#issuecomment-412182628)

Or are you asking for somethig else?

**Comment by nukec** [comment]
@halter73 i resolved this by changing 32bit to 64bit app in azure for service app.

**Comment by halter73** [comment]
@ufalak The reason I ask is because you have two different exceptions in this issue. The exceptions in your first comment ("IOException: ... InnerException: SocketException (10054): An existing connection was forcibly closed by the remote host") was thrown during JsonInputFormatter.ReadRequestBodyAsync()'s call to Request.Body.ReadAsync(). Throwing for socket errors that occur while reading data from the client the expected behavior for Kestrel.

The exception message I'm interested in now is the one from your most recent comment which was thrown during StaticFileContext.SendAsync()'s call to Response.Body.WriteAsync(). If this was also caused by the same type of SocketException (which I don't see how it could be), this would be a bug in Kestrel because writing the the response body stream after the client prematurely disconnects is not supposed to raise an exception through Response.Body.WriteAsync() unless the RequestAborted CancellationToken is passed into the write call, but in that case you would get an OperationCanceledException instead of an IOException. Without a CancellationToken, the data written to the response body pipe basically goes to /dev/null since, as you mention, throwing an exception in this case would usually be useless.

So, long story short, if you can get the full exception message and stack trace together for the exception thrown from StaticFileContext.SendAsync(), that could be helpful to determine what exception we're not handling, because it can't be the same "SocketException (10054): ..." you saw from JsonInputFormatter.ReadRequestBodyAsync().

**Comment by ufalak** [comment]
@halter73 it wasnt me the one who write that second exception.
It was @nukec.

**Comment by halter73** [comment]
Sorry for my confusion. @nukec Can you provide the exception message for the IOException you posted?

**Comment by AlanMacdonald** [comment]
We're getting this error thousands of times a day in our 10 server web farm now we have gone out to production as of 2 days ago with a new site for processing user logins. It does not seem to correspond to a visible issue for the user.  Can someone please confirm we're saying this is safe to be occurring and so really we should be looking at catching and suppressing the logging of this error?

**Comment by halter73** [comment]
@AlanMacdonald "SocketException (10054): An existing connection was forcibly closed by the remote host" simply means that a client disconnected ungracefully. This can happen if the client process exited mid-connection for example.

Usually this doesn't cause any issues visible to the user, since the user typically isn't concerned what happens on the server after a client closes the connection. That's why Kestrel logs this with a relatively low "INFO" severity. If the client disconnects in the middle of uploading a request body, that can cause HttpContext.Request.Body.ReadAsync() to throw leading to higher severity logs, but that's pretty rare.

**Comment by benaadams** [comment]
@AlanMacdonald do you know what kind of devices you users are on? You'll see more ungraceful disconnects for mobile clients than for desktops for example.

**Comment by AlanMacdonald** [comment]
@halter73 thanks for the info.

@benaadams it's mainly desktop users but the login site where this is occurring is identityserver4 so there could be things happening internal there maybe or it could be something with our backchannel communication between main site and login site but that all seems to be hanging together. Thanks

**Comment by KholodovOleksandr** [comment]
we have the same error. our environment is azure VMs with a load balancer and 2+ machines. it reproduced time to time in firefox while image uploading
